### PR TITLE
feat(samcli): support `sam sync`

### PR DESCRIPTION
**Pull request by JadenSimon**
## Problem
No support for `sam sync`

## Solution
* Implement logic to run `sam sync` with a corresponding wizard
    * This wizard is similar to the legacy flow except it does not ask for parameter overrides
* Add setting to go back to legacy `sam deploy` flow

### TODO
* ~Add tests~
* ~Do not use `vscode.Pseudoterminal` on C9~
* ~Add experiment for "code" syncs~
* ~Bucket prompt should have an option to let SAM CLI automatically create a bucket~

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -138,6 +138,11 @@
                     "default": 90000,
                     "markdownDescription": "%AWS.configuration.description.samcli.lambdaTimeout%"
                 },
+                "aws.samcli.enableLegacyDeploy": {
```
Comment (justinmk3) [inline package.json:141]: nit: usually don't need "enable" in settings names, it is implied
```suggestion
                "aws.samcli.legacyDeploy": {
```

File: src/shared/utilities/childProcess.ts [hunk 2]
```diff
@@ -308,6 +308,24 @@ export class ChildProcess {
...
+     * This throws if the process hasn't started or if the write fails.
+     */
+    public async send(input: string) {
+        if (this.childProcess === undefined) {
+            throw new Error('Cannot write to non-existent process')
+        }
+
+        const stdin = this.childProcess.stdin
+        if (!stdin) {
+            throw new Error('Cannot write to non-existent writable')
```
Comment (justinmk3) [inline src/shared/utilities/childProcess.ts:323]: message seems a bit cryptic

File: src/shared/utilities/collectionUtils.ts [hunk 3]
```diff
@@ -342,3 +342,23 @@ export function pageableToCollection<
 
     return toCollection(gen)
 }
+
+/**
+ * Converts an iterable of promises into an unordered stream of values.
+ *
+ * The resulting stream will throw if any of the promises are rejected.
+ */
+export async function* toStream<T>(values: Iterable<Promise<T>>): AsyncGenerator<T, void> {
```
Comment (justinmk3) [inline src/shared/utilities/collectionUtils.ts:351]: cool!

File: src/shared/sam/sync.ts [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+
+    if (resp === okDontShow) {
+        await PromptSettings.instance.disablePrompt('samcliConfirmDevStack')
+    }
+}
+
+// This is a decent improvement over using the output channel but it isn't a tty/pty
+// SAM CLI uses `click` which has reduced functionality if `os.isatty` returns false
+// Historically, Windows lack of a pty-equivalent is why it's not available in libuv
+// Maybe it's doable now with the ConPTY API? https://github.com/libuv/libuv/issues/2640
```
Comment (justinmk3) [inline src/shared/sam/sync.ts:363]: yes, should be possible in latest libuv/nodejs with Windows Terminal.

**Review by justinmk3 — APPROVED** [review]
changelog?

**Comment by JadenSimon** [comment]
Need to work on the changelog. It's hard to write out all the improvements succinctly. 

>  Error: exposeEmitters(): failed to find emitters for keys "onDidAccept", "onDidTriggerButton"

RIP. They may be moving to true private fields.

File: src/shared/sam/sync.ts [hunk 5]
```diff
@@ -211,16 +215,20 @@ export async function runSamSync(args?: Partial<SyncParams>) {
         ecrRepoUri: '--image-repository',
     })
 
-    // TODO: use `SamCliContext` for the executable path
-    const sam = new ChildProcess('sam', ['sync', ...params], {
+    const { path: samCliPath } = await SamCliSettings.instance.getOrDetectSamCli()
+    if (samCliPath === undefined) {
+        throw new ToolkitError('A location for SAM CLI could not be found', { code: 'MissingExecutable' })
```
Comment (justinmk3) [inline src/shared/sam/sync.ts:220]:
```suggestion
        throw new ToolkitError('SAM CLI could not be found', { code: 'MissingExecutable' })
```

File: src/test/shared/sam/sync.test.ts [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+        it('can load a relative template param', async function () {
+            const template = await makeTemplateItem(tempDir)
+            const params = await getParams(`template = "./template.yaml"`)
+            assert.deepStrictEqual(params.template?.data, template.data)
+        })
+
+        it('can load an absolute template param', async function () {
+            const template = await makeTemplateItem(tempDir)
+            const params = await getParams(`template = "${template.uri.fsPath}"`)
```
Comment (JadenSimon) [inline src/test/shared/sam/sync.test.ts:133]:
```suggestion
            const params = await getParams(`template = '${template.uri.fsPath}'`)
```

File: src/integrationTest/sam.test.ts [hunk 7]
```diff
@@ -397,6 +419,58 @@ describe('SAM Integration Tests', async function () {
         console.log(`DebugSessions seen in this run:\n${sessionReport}`)
     })
 
+    describe('sync', function () {
```
Comment (JadenSimon) [inline src/integrationTest/sam.test.ts:422]: TODO: move this to `sync.test.ts` and skip it when `process.env['AWS_TOOLKIT_AUTOMATION'] !== 'integration'`

Having a bunch of separate files for integration tests is unnecessary IMO.

Comment (JadenSimon) [reply src/integrationTest/sam.test.ts:422]: Also the clean-up logic I added could be improved. 404s should be treated as a 'Success' and other errors should throw. Resources that failed to delete should be persisted and retried later.

Comment (JadenSimon) [reply src/integrationTest/sam.test.ts]: This needs infra changes.

**Review by justinmk3 — APPROVED** [review]

**Review by jpinkney-aws — APPROVED** [review]

**Pull request merged**